air/src/execution_step/instructions/canon.rs

```diff
@@ -14,189 +14,80 @@
  * limitations under the License.
  */
 
+use super::canon_utils::handle_seen_canon;
+use super::canon_utils::handle_unseen_canon;
+use super::canon_utils::CanonEpilogClosure;
+use super::canon_utils::CreateCanonStreamClosure;
 use super::ExecutionCtx;
 use super::ExecutionResult;
 use super::TraceHandler;
 use crate::execution_step::boxed_value::CanonStream;
 use crate::execution_step::boxed_value::CanonStreamWithProvenance;
-use crate::execution_step::instructions::resolve_peer_id_to_string;
-use crate::execution_step::Stream;
+use crate::execution_step::Generation;
 use crate::log_instruction;
 use crate::trace_to_exec_err;
-use crate::UncatchableError;
 
 use air_interpreter_cid::CID;
-use air_interpreter_data::CanonCidAggregate;
 use air_interpreter_data::CanonResult;
 use air_interpreter_data::CanonResultCidAggregate;
 use air_parser::ast;
-use air_parser::ast::ResolvableToPeerIdVariable;
 use air_parser::AirPos;
 use air_trace_handler::merger::MergerCanonResult;
 
 use std::borrow::Cow;
 use std::rc::Rc;
 
-pub(super) type CanonEpilogClosure<'ctx> =
-    dyn for<'i> Fn(StreamWithSerializedView, &mut ExecutionCtx<'i>, &mut TraceHandler) -> ExecutionResult<()> + 'ctx;
-
 impl<'i> super::ExecutableInstruction<'i> for ast::Canon<'i> {
     #[tracing::instrument(level = "debug", skip(exec_ctx, trace_ctx))]
     fn execute(&self, exec_ctx: &mut ExecutionCtx<'i>, trace_ctx: &mut TraceHandler) -> ExecutionResult<()> {
         log_instruction!(call, exec_ctx, trace_ctx);
+        let epilog = &epilog_closure(self.canon_stream.name);
         let canon_result = trace_to_exec_err!(trace_ctx.meet_canon_start(), self)?;
 
-        let epilog: &CanonEpilogClosure<'_> = &|stream_with_positions: StreamWithSerializedView,
-                                                exec_ctx: &mut ExecutionCtx<'_>,
-                                                trace_ctx: &mut TraceHandler|
-         -> ExecutionResult<()> {
-            let StreamWithSerializedView {
-                canon_stream,
-                canon_result_cid,
-            } = stream_with_positions;
-
-            let value = CanonStreamWithProvenance::new(canon_stream, canon_result_cid.clone());
-            exec_ctx.scalars.set_canon_value(self.canon_stream.name, value)?;
-
-            trace_ctx.meet_canon_end(CanonResult::new(canon_result_cid));
-            Ok(())
-        };
-
         match canon_result {
             MergerCanonResult::CanonResult(canon_result_cid) => {
                 handle_seen_canon(epilog, canon_result_cid, exec_ctx, trace_ctx)
             }
             MergerCanonResult::Empty => {
-                let get_stream_or_default = get_stream_or_default_closure(self.stream.name, self.stream.position);
-                handle_unseen_canon(epilog, &get_stream_or_default, &self.peer_id, exec_ctx, trace_ctx)
+                let create_canon_stream = create_canon_stream_closure(self.stream.name, self.stream.position);
```

Comment:
Technically create_canon_stream is the closure so I couldn't get from the naming what is what.
```suggestion
                let canon_stream_producer = create_canon_stream_producer(self.stream.name, self.stream.position);
```